### PR TITLE
Add Architecture Guide

### PR DESCRIPTION
**Pull request by davish**
Follow up to #303 that adds an architecture guide to help future contributors.

### CODE REVIEW

File: README.md
Comment (davish) [inline README.md:1]: this file looks exactly the same

File: src/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+Translation between `OFCEvent` and `EventInput` is handled in `interop.ts`. Each `Calendar` subclass handles translation between its source format and `OFCEvent`.
+
+Objects can be validated as OFCEvents using `validateEvent()` . This function is used throughout the code to ensure that only valid events are present.
+
+### `core`
+
+The `core` directory consists of two classes, `EventStore` and `EventCache`. These two classes comprise the plugin's main event-managing logic.
+
+The `EventStore` is the source of truth for events in the plugin. Its interface is similar to a simplified database that stores events, calendars and file locations. Files and calendars are one-to-many relationships: every event is related to a calendar and at most one file, but calendars and files can have many events within them. The `EventStore` allows for effecient querying of events grouped by calendars and files. Every event in the `EventStore` has an ID associated with it. Local events have random IDs that are generated at insert time, but remote events using the iCal spec have `UID`s that are plumbed through.
```
Comment (KeenWill) [inline src/README.md:60]: can an event have no file associated with it? or is it always exactly one?

File: src/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+## Codemap
+
+Following the advice in [this blog post on architecture docs](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html), the following section will list out the main modules of the code without linking out to specific file locations that may quickly become stale. Make use of code search and the TypeScript Language Server to jump around and explore the code, with this section as your guide.
+
+### `types`
+
+This module defines some common types used throughout the code. The most prevalent is `OFCEvent`, short for Obsidian Full Calendar Event, that specifies the intermediate representation for all events in the plugin. Note that FullCalendar.io uses a different event format called `EventInput`, which you can read about [in their documentation](https://fullcalendar.io/docs/event-parsing).
+
+Translation between `OFCEvent` and `EventInput` is handled in `interop.ts`. Each `Calendar` subclass handles translation between its source format and `OFCEvent`.
```
Comment (KeenWill) [inline src/README.md:52]: might be good to add one sentence here explaining what a Calendar type does / represents

File: src/README.md (hunk 1 above)
Comment (davish) [reply src/README.md:60]: Clarified that every event has `exactly one` calendar, and `at most one` file. Remote events don't have a file associated with them, yet.

File: src/README.md (hunk 2 above)
Comment (davish) [reply src/README.md:52]: add a `(see below)` label.

File: src/README.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+Obsidian Full Calendar's goal is to give users a robust and feature-ful calendar view into their Vault. In addition to displaying and modifying events stored in note frontmatter and daily note bulleted lists, it can also read events from the Internet in CalDAV and ICS format.
+
+Obsidian Full Calendar takes its name from [FullCalendar](https://github.com/fullcalendar/fullcalendar), a "Full-sized drag & drop event calendar in JavaScript." This plugin uses FullCalendar as its calendar view. While the naming can be ambiguous, this document will always refer to the FullCalendar view library without any spaces, or as `fullcalendar.io`. The plugin will be referred to either as "the plugin", "Full Calendar" with a space, or "Obsidian Full Calendar".
+
+As of now, the plugin supports events from the following sources and formats:
+
+-   Frontmatter of notes in the open Obsidian Vault.
+-   Bullet list items in Daily Notes generated by the [core Daily Notes](https://help.obsidian.md/Plugins/Daily+notes) or [Periodic Notes](https://github.com/liamcain/obsidian-periodic-notes) plugins.
+-   ICS files served on the Internet.
```
Comment (pawalt) [inline src/README.md:11]: Can you clarify what "served on the internet" means - in particular in distinction to how CalDAV works?

File: src/README.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+`EditableCalendar`s are constructed with references to an `ObsidianAdapter` instance that handles all interactions with the Obsidian API. This adapter is useful for testing, since it reduces the surface area of APIs to be mocked from the entire API to a handful of functions that the plugin actually uses.
+
+### `ui`
+
+While `core` and `calendars` make up the Model in the `MVC` pattern, the Views and Controllers are currently both living in the `ui` directory. The view connector to the FullCalendar library lives in `calendar.ts`. Most of the controller logic that interfaces with the `EventCache` lives, somewhat confusingly, in `view.ts`. Auxilliary views, like the edit/create modal and settings selectors, are React components that live in their own `.tsx` files and are mounted into the DOM when needed.
+
+**Architecture Invariant**: All interactions with event data should be mediated by the `EventCache`. Code in the `ui` directory should not reference or call out to the `EventStore`, Obsidian Vault APIs, or `Calendar` subclasses.
```
Comment (pawalt) [inline src/README.md:84]: Can this be enforced with a linter?

Comment (davish) [reply src/README.md:84]: *maybe*? It's enforced at the code review level right now. I don't think it's worth it to figure out a linter rule for this.

File: src/README.md (hunk 3 above)
Comment (davish) [reply src/README.md:11]: rephrased to be `ICS files publicly accessible at a URL` to draw a clearer distinction with CalDAV servers behind basic auth, and also to clarify that it's not just the internet -- localhost URLs work

**Pull request merged**